compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -820,10 +820,29 @@ bool OMR::CFGSimplifier::simplifyBooleanStore(bool needToDuplicateTree)
       {
       TR::Node *base1 = store1->getFirstChild();
       TR::Node *base2 = store2->getFirstChild();
-      if (!base1->getOpCode().hasSymbolReference() || !base2->getOpCode().hasSymbolReference())
-         return false;
-      if (base1->getSymbolReference()->getReferenceNumber() != base2->getSymbolReference()->getReferenceNumber())
+      // accessing the same symbol
+      if (base1->getOpCode().hasSymbolReference() && base2->getOpCode().hasSymbolReference())
+         {
+         if (base1->getSymbolReference()->getReferenceNumber() != base2->getSymbolReference()->getReferenceNumber())
+            return false;
+         }
+      // offsetting the same amount from the same symbol
+      else if ((base1->getOpCodeValue() == TR::aladd && base2->getOpCodeValue() == TR::aladd)
+               || (base1->getOpCodeValue() == TR::aiadd && base2->getOpCodeValue() == TR::aiadd))
+         {
+         if (!base1->getFirstChild()->getOpCode().hasSymbolReference()
```

Comment:
Does this condition also need to check if the base's first children are not indirect loads or calls (meaning more testing is required to ensure that we are in fact storing to the same indirect store location target) ?

Reply:
I'm not sure I agree - the situation we have here is a diamond with a single store on each side. We are checking the store to see if we can do a single of a condition test. The code decided this was safe to do for all store indirects if the store indirects were storing to the same symref and the base object was from a node which had a symref and the two sides of the diamond had the same symref. I am generalizing that logic to the just allow a constant offset from that same base object. If I need to further limit what base1 and base2 are then the original code also has a bug.

Reply:
The case I'm asking about is:

```
if (cond)
   o1.f[2].g = value1;
else
   o2.f[2].g = value2;

```
with respect to your change (where o1 != o2). If your point is that a different case handled by the original code that looks like:

```
if (cond)
   o1.f.g = value1;
else
   o2.f.g = value2;

```
is also problematic, then I don't disagree with that (i.e. I was raising a question about the code that I reviewed) but that does not take away from the fact that there is a change that may be needed to fix a potential bug in this area of the code.

Reply:
Ok yup agreed - I wasn't sure if it was a comment directed at the new implementation or the implementation as a whole. It is probably a cross-cutting concern (for both the old and new implementations so I'll have a look and see if I can put in a suitable check.